### PR TITLE
8298865: Excessive memory allocation in CipherOutputStream AEAD decryption

### PR DESCRIPTION
**Pull request by djelinski**
This patch modifies `CipherOutputStream` to avoid pointless memory allocations when decrypting data using AEAD ciphers. This is related to #11597, which fixed a similar issue in `CipherInputStream`.

`Cipher.update` does not output any data when doing AEAD decryption; all data is buffered, and is later returned in one shot from `doFinal`. `Cipher.getOutputSize` returns the buffer size required by `doFinal`, which increases after every `update`, triggering new allocation in `ensureCapacity`.

This patch addresses the issue by calling the `update` overload that returns the output buffer until one of the update calls returns some data. When that happens, we know that the cipher does not buffer everything until `doFinal`, and revert to original behavior.

This PR adds a new benchmark for AES/GCM encryption and decryption using `CipherOutputStream`.

Benchmark results before:
```
Benchmark                         (dataSize)  (keyLength)  (provider)   Mode  Cnt       Score      Error  Units
AESGCMCipherOutputStream.decrypt       16384          128              thrpt   40   27949,624 ±  301,408  ops/s
AESGCMCipherOutputStream.decrypt     1048576          128              thrpt   40      20,730 ±    0,875  ops/s
AESGCMCipherOutputStream.encrypt       16384          128              thrpt   40  175358,641 ± 4235,808  ops/s
AESGCMCipherOutputStream.encrypt     1048576          128              thrpt   40    2588,111 ±   35,469  ops/s
```

after:
```
Benchmark                         (dataSize)  (keyLength)  (provider)   Mode  Cnt       Score      Error  Units
AESGCMCipherOutputStream.decrypt       16384          128              thrpt   40   69644,217 ± 1081,032  ops/s
AESGCMCipherOutputStream.decrypt     1048576          128              thrpt   40     949,667 ±    9,431  ops/s
AESGCMCipherOutputStream.encrypt       16384          128              thrpt   40  173144,038 ± 3279,149  ops/s
AESGCMCipherOutputStream.encrypt     1048576          128              thrpt   40    2514,840 ±   87,935  ops/s
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298865](https://bugs.openjdk.org/browse/JDK-8298865): Excessive memory allocation in CipherOutputStream AEAD decryption

### Reviewers
 * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**) ⚠️ Review applies to [7b295132](https://git.openjdk.org/jdk/pull/11693/files/7b295132a7e594a424feb220eeb222adb3002b63)
 * [Anthony Scarpino](https://openjdk.org/census#ascarpino) (@ascarpino - **Reviewer**) ⚠️ Review applies to [7b295132](https://git.openjdk.org/jdk/pull/11693/files/7b295132a7e594a424feb220eeb222adb3002b63)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11693/head:pull/11693` \
`$ git checkout pull/11693`

Update a local copy of the PR: \
`$ git checkout pull/11693` \
`$ git pull https://git.openjdk.org/jdk pull/11693/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11693`

View PR using the GUI difftool: \
`$ git pr show -t 11693`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11693.diff">https://git.openjdk.org/jdk/pull/11693.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/javax/crypto/CipherOutputStream.java [hunk 1]
```diff
@@ -91,11 +91,17 @@ public class CipherOutputStream extends FilterOutputStream {
      * Ensure obuffer is big enough for the next update or doFinal
      * operation, given the input length {@code inLen} (in bytes)
      *
+     * If obuffer is null/zero-sized, do not allocate a new buffer.
+     * This reduces allocation for AEAD ciphers that never return data from update
```
Comment (valeriepeng) [inline src/java.base/share/classes/javax/crypto/CipherOutputStream.java:95]: AEAD ciphers do return data for update() calls for encryption. Perhaps we should add "when used for decryption" or some other similar wordings to the above sentence? Same goes for the comment in CipherInputStream class.

**Review by valeriepeng — APPROVED** [review]

**Review by ascarpino — APPROVED** [review]

File: src/java.base/share/classes/javax/crypto/CipherOutputStream.java (hunk 1 above)
Comment (ascarpino) [reply src/java.base/share/classes/javax/crypto/CipherOutputStream.java:95]: Perhaps it's the way you read the sentence, when I read it bit before your comment, I interpreted the change as an open-ended comment where AEAD may or may not return without being specific.  I'm neutral to changing it to specify encryption vs decryption.

Comment (djelinski) [reply src/java.base/share/classes/javax/crypto/CipherOutputStream.java]: That's a good suggestion; I changed the wording now. What do you think?

Comment (valeriepeng) [reply src/java.base/share/classes/javax/crypto/CipherOutputStream.java]: Look good, thanks!

**Pull request closed without merging**